proposal/nested-map-functions.md

```diff
@@ -0,0 +1,204 @@
+`$map2` should be merged. For example, we update the hover colors in our `$nav`
+map above:
+
+```scss
+@use 'sass:map';
+
+$new-hover: (
+  'search': green,
+  'logo': orange,
+);
+
+$nav: map.merge($nav, 'color', 'hover', $new-hover);
+
+// $nav: (
+//   'bg': 'gray',
+//   'color': (
+//     'hover': (
+//       'search': green,
+//       'home': red,
+//       'filter': blue,
+//       'logo': orange,
+//     ),
+//   ),
+// );
+```
+
+This proposal also adds a `set()` function to `sass:map`, with a similar syntax,
+returning a map with any nested key set to a specific value. To achieve the
+same output as our merge example, we can set each key individually:
+
+```scss
+@use 'sass:map';
+
+$nav: map.set($nav, 'color', 'hover', 'search', green);
+$nav: map.set($nav, 'color', 'hover', 'logo', orange);
+```
+
+## Syntax
+
+### `get()`
+
+```
+get($map, $keys...)
+```
+
+* Let `key` be the first (or only) element in `$keys`
+
+* If `$map` does not have a key with the same name as `key`, throw an error.
+
+* Let `value` be the value assigned to `key` in `$map`
+
+* If there is more than one element in `$keys`:
+
+  * Let `keys` be all elements in `$keys` after the first element
+
+  * Call `get()` with `value` and expanded `keys` as arguments
+
+* Otherwise, return `value`
+
+### `has-key()`
+
+```
+has-key($map, $keys...)
+```
+
+* Let `key` be the first (or only) element in `$keys`
+
+* If `$map` does not have a key with the same name as `key`, return boolean
+  `false`.
+
+* If there is more than one element in `$keys`:
+
+  * Let `map` be the value assigned to `key` in `$map`
+
+  * Let `keys` be all but the first element in `$keys`
+
+  * Call `has-key()` with `map` and expanded `keys` as arguments
+
+* Otherwise, return boolean `true`
+
+### `set()`
+
+```
+set($map1, $keys..., $value)
+```
+
+* If `$map1` is not a map, throw an error.
+
+* If fewer than three arguments are provided, throw an error.
+
+* Let `map` be an empty map .
+
+* If there is more than one argument in arglist `$keys`:
+
+  * Let `set-key` be the first element in `$keys`.
+
+  * Let `keys` be a slice of all `$keys` elements except the first.
+
+  * If `$map1` has a key `current-key` with the same name as `set-key`:
+
+    * Let `current` be the value of `current-key`.
+
+  * Otherwise:
+
+    * Let `current` be an empty map.
+
+  * Let `set-value` be the result of calling `set()` with `current` and expanded
+      `keys` as arguments.
+
+    > This will error if `current` is not a map, but we stil have nested `keys`
+
+* Otherwise:
+
+  * Let `set-key` be the only element in `$keys`
+
+  * Let `set-value` be the value of `$value`
+
+* Add a key with name `set-key` and value `set-value` to `map`
+
+* For each `key`, `value` pair in `$map1`
+
+  * If a key with name `key` already exists in `map`, do nothing.
+
+  * Otherwise, add a key to `map` with name `key` and value `value`.
+
+* Return the value `map`
+
+### `merge()`
+
+```
+merge($map1, $keys..., $map2)
```

Comment:
Similar comment to `map.set()`, with the caveat that we *do* need to continue to support `map.merge($map1: ..., $map2: ...)` for backwards-compatibility. You can get away with this by just saying that this is adding a new overload, like I did in https://github.com/sass/sass/blob/master/accepted/color-4-rgb-hsl.md.